### DISCUSSION TITLE
How can I get a trajectory interpolated with a specific dt?

### DISCUSSION THREAD

**Discussion by blooop** [NVlabs/curobo#210] [Software Q&A] [answered]
I am calling plan_single multiple times and combining the results into a single trajectory.  I noticed that the motion gen result has different values for optimize_dt, which makes it harder to combine into a single unified trajectory before execution.  I would also like to be able to run the controller at a fixed timestep.  

I looked into turning optimize_dt off, but that seems like it has other considerations as mentioned here:
https://github.com/NVlabs/curobo/discussions/78

and I found I got weird results if dt<=0.01.  It also doesn't seem like the correct approach as it forces the optimisation to run at a higher resolution than required.

I found:
https://curobo.org/_api/curobo.graph.graph_base.html#curobo.graph.graph_base.GraphPlanBase.get_interpolated_trajectory

but have not been able to get it to get it working properly because of the initialisation of out_traj_state.

```
    inter_plan = result.get_interpolated_plan()
    out_inter_plan = inter_plan.clone()
    curobo.util.trajectory.get_interpolated_trajectory(
        inter_plan, out_inter_plan, interpolation_dt=0.02
    )
```  
Passing in an empty joint state doesn't seem to work either.  Is this the right approach or is there a better way of getting the result of plans with a fixed dt?

Thanks

**Comment by balakumar-s** [comment]
At what dt do you want the interpolated trajectory to be? We have had success with `interpolation_dt=0.004` (250hz). 

Here is one way to stitch together trajectories (which we use) :
1. Create motion gen instance with required interpolation_dt: https://github.com/NVlabs/curobo/blob/774dcfd609847ad98147f166ea2f5954516ff7a0/src/curobo/wrap/reacher/motion_gen.py#L157
2. After every planning query, call `get_interpolated_plan()` from MotionGenResult to get interpolated trajectory: https://github.com/NVlabs/curobo/blob/774dcfd609847ad98147f166ea2f5954516ff7a0/src/curobo/wrap/reacher/motion_gen.py#L806
3. Clone this interpolated_plan and store in a list.

Then you will have all trajectories.

**Reply by blooop** [reply]
Thanks, your approach is the first thing I tried, but I needed to set optimize_dt=False which I from the reports sounds like it will result in worse plans, and cause issues with tuning as you mention here: 

https://github.com/NVlabs/curobo/discussions/78

I was hoping to be able to chose the dt of the interpolated trajectory after the fact.  So I can leave optimize_dt=True and then take the resultant trajectory and sample it at any arbitrary dt.

**Reply by balakumar-s** [reply, marked as answer]
Why do you need to set optimize_dt to False. get_interpolated_plan() will always give you a fixed dt trajectory. The fixed dt is obtained from interpolation_dt.

**Reply by blooop** [reply]
Ah, I think this explains my confusion.  I based my code off the motion planning example which had as the last line:
`print("Trajectory Generated: ", result.success, result.optimized_dt.item())`

and from there I was using optimised dt as the dt to execute the plans.  

It sounds like the curobo is already returning the trajectory with the specified interpolation_dt, so there is no need for me to resample it after. 

Sorry for the massive confusion. I think we were talking at cross purposes most of the time, because of my fundamental incorrect assumption!

Thanks

**Comment by scharalambous-sony** [comment]
I believe you should also turn optimize_dt off, or otherwise the finetuning stage will optimize the dt (to push the trajectory to the vel/acc/jerk limits)

**Reply by blooop** [reply]
From my understanding of the report, that is not what optimise_dt does. It iteratively chooses the optimal timestep in which to perform the trajectory optimisation.  Even if optimise_dt=False the trajectory will still try to push the trajectory towards the vel/acc/jerk limits, but because the dt might be too large or too small, that optimisation process might either take too long or not have enough fidelity for the complexity of the trajectory.  (I might have misunderstood this though hence starting this discussion)

I would like the benefits of optimise_dt=True but then be able to resample the final trajectory at an arbitrary timestep.

**Reply by scharalambous-sony** [reply]
If optimize_dt is False the retiming does not take place:
https://github.com/NVlabs/curobo/blob/774dcfd609847ad98147f166ea2f5954516ff7a0/src/curobo/util/trajectory.py#L551

For interpolation you may be able to use a function from[ src/curobo/util/trajectory.py](https://github.com/NVlabs/curobo/blob/774dcfd609847ad98147f166ea2f5954516ff7a0/src/curobo/util/trajectory.py#L551) but I am not sure if there is any interpolation util there that is decoupled from the retiming

**Reply by blooop** [reply]
Thanks, yeah.  I was looking into using:
https://github.com/NVlabs/curobo/blob/774dcfd609847ad98147f166ea2f5954516ff7a0/src/curobo/util/trajectory.py#L306 
from that page, but was not sure how to initialise the out_traj_state variable.

**Reply by scharalambous-sony** [reply]
You could use something like this
https://github.com/NVlabs/curobo/blob/774dcfd609847ad98147f166ea2f5954516ff7a0/src/curobo/util/trajectory.py#L157

**Comment by scharalambous-sony** [comment]
@blooop @balakumar-s I have created a pull request that resolved similar issues that I was also facing with time optimization https://github.com/NVlabs/curobo/pull/217

**Reply by balakumar-s** [reply]
Thanks for the PR. The one limitation we have currently is that `min_dt` cannot be lower than `interpolation_dt`. If `min_dt` is lower, then the interpolation function will try to interpolate from 32 timesteps at `min_dt` to a <32 timesteps with  `interpolation_dt` which can cause jumps in the interpolated trajectory. We have to also implement a better interpolation function to fix this.

Can you describe your use case where the current implementation fails? We would like to write a unit test for this.

**Reply by scharalambous-sony** [reply]
I understand, thanks for the information. My motivation in decoupling minimum from interpolation dt is to make the time optimization interpetable as it was confusing why they were coupled. I'd recommend decoupling them, and raising a warning when min_dt is set lower than interpolation_dt

The current implementation fails when the minimum dt is a low number (e.g. 0.01). The failure occurs in compute_smoothness. Intuitively, we are retiming the trajectory to exactly the vel/acc/jerk limits. Any numerical error will cause this check to fail. Hence, we can either retime to the limits tightened by a tolerance, or relax the check in compute_smoothness by a tolerance. I have chosen the latter

**Reply by scharalambous-sony** [reply]
This is possibly related to #222
